From: Andreas Herrman <aherrman@de.ibm.com>
Subject: Failed pathes remain unavailable
References: 173872 - LTC23720

When a port becomes unavailable the zfcp driver should notify
the midlayer about the new port status with fc_rport_delete().

Signed-off-by: Hannes Reinecke <hare@suse.de>

diff -Nurp -X /home/aherrman/ignore-diff linux-2.6.16.11-9-orig/drivers/s390/scsi/zfcp_erp.c linux-2.6.16.11-9/drivers/s390/scsi/zfcp_erp.c
--- linux-2.6.16.11-9-orig/drivers/s390/scsi/zfcp_erp.c	2006-05-11 11:39:33.000000000 +0200
+++ linux-2.6.16.11-9/drivers/s390/scsi/zfcp_erp.c	2006-05-11 17:19:07.000000000 +0200
@@ -3407,10 +3407,12 @@ zfcp_erp_action_cleanup(int action, stru
 		break;
 	case ZFCP_ERP_ACTION_REOPEN_PORT_FORCED:
 	case ZFCP_ERP_ACTION_REOPEN_PORT:
-		if ((result == ZFCP_ERP_SUCCEEDED)
-		    && !atomic_test_mask(ZFCP_STATUS_PORT_NO_WWPN,
-					 &port->status)
-		    && !port->rport) {
+		if (atomic_test_mask(ZFCP_STATUS_PORT_NO_WWPN, &port->status)) {
+			zfcp_port_put(port);
+			break;
+		}
+		
+		if ((result == ZFCP_ERP_SUCCEEDED) && !port->rport) {
 			struct fc_rport_identifiers ids;
 			ids.node_name = port->wwnn;
 			ids.port_name = port->wwpn;
@@ -3426,9 +3428,23 @@ zfcp_erp_action_cleanup(int action, stru
 			else
 				scsi_flush_work(adapter->scsi_host);
 		}
+		if ((result != ZFCP_ERP_SUCCEEDED) && port->rport) {
+			fc_remote_port_delete(port->rport);
+			port->rport = NULL;
+		}
 		zfcp_port_put(port);
 		break;
 	case ZFCP_ERP_ACTION_REOPEN_ADAPTER:
+		if (result != ZFCP_ERP_SUCCEEDED) {
+			struct zfcp_port *port;
+			list_for_each_entry(port, &adapter->port_list_head, list)
+				if (port->rport &&
+				    !atomic_test_mask(ZFCP_STATUS_PORT_WKA,
+						      &port->status)) {
+					fc_remote_port_delete(port->rport);
+					port->rport = NULL;
+				}
+		}
 		zfcp_adapter_put(adapter);
 		break;
 	default:
diff -Nurp -X /home/aherrman/ignore-diff linux-2.6.16.11-9-orig/drivers/s390/scsi/zfcp_scsi.c linux-2.6.16.11-9/drivers/s390/scsi/zfcp_scsi.c
--- linux-2.6.16.11-9-orig/drivers/s390/scsi/zfcp_scsi.c	2006-05-11 11:39:33.000000000 +0200
+++ linux-2.6.16.11-9/drivers/s390/scsi/zfcp_scsi.c	2006-05-11 11:35:24.000000000 +0200
@@ -877,11 +877,17 @@ zfcp_reset_fc_host_stats(struct Scsi_Hos
 	}
 }
 
+static void zfcp_set_rport_dev_loss_tmo(struct fc_rport *rport, u32 timeout)
+{
+	rport->dev_loss_tmo = timeout;
+}
+
 struct fc_function_template zfcp_transport_functions = {
 	.show_starget_port_id = 1,
 	.show_starget_port_name = 1,
 	.show_starget_node_name = 1,
 	.show_rport_supported_classes = 1,
+	.show_rport_dev_loss_tmo = 1,
 	.show_host_node_name = 1,
 	.show_host_port_name = 1,
 	.show_host_permanent_port_name = 1,
@@ -889,6 +895,7 @@ struct fc_function_template zfcp_transpo
 	.show_host_supported_speeds = 1,
 	.show_host_maxframe_size = 1,
 	.show_host_serial_number = 1,
+	.set_rport_dev_loss_tmo = zfcp_set_rport_dev_loss_tmo,
 	.get_fc_host_stats = zfcp_get_fc_host_stats,
 	.reset_fc_host_stats = zfcp_reset_fc_host_stats,
 	/* no functions registered for following dynamic attributes but
